### ISSUE TITLE
[Research] токены onDark / onLight / Inverse

### ISSUE THREAD

**Issue by neretin-trike** [salute-developers/plasma#1152]
[Ссылка](https://codesandbox.io/p/sandbox/plasma-web-example-forked-l6jwsc?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu9ppflr00063b6hyhghw9kj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu9ppflr00023b6h1oqv3zx0%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu9ppflr00033b6hsosp0sk6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu9ppflr00053b6hbvccv2up%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu9ppflr00023b6h1oqv3zx0%2522%253A%257B%2522id%2522%253A%2522clu9ppflr00023b6h1oqv3zx0%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clu9ppflr00053b6hbvccv2up%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu9ppflr00043b6htwcu10wb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu9ppflr00053b6hbvccv2up%2522%252C%2522activeTabId%2522%253A%2522clu9ppflr00043b6htwcu10wb%2522%257D%252C%2522clu9ppflr00033b6hsosp0sk6%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu9ppflr00033b6hsosp0sk6%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) на прототип подхода.

### Как использовать
1. Обернуть нужные компоненты в контейнер (или использовать существующий)
2. Импортировать applyOnLight / applyOnDark / applyInverse из plasma-web / plasma-b2c и т.д.
3. Применить нужный миксин внутри контейнера и выбрать компоненты, которые нужно стилизовать (по дефолту стилизуются все компоненты)

### Спецификация
- applyOnLight - применяется всегда на светлом фоне независимо от выбранного режима темы (dark / light)
- applyOnDark - применяется всегда на тёмном фоне независимо от выбранного режима темы (dark / light)
- applyInverse - применяется для фонов, которые меняются в зависимости от выбранного режима темы (dark / light)

### Шаги для реализации такого подхода
1. Провести ревизию всех токенов цветов в конфигах  компонент для библиотек plasma-web / plasma-b2c / plasma-asdk
2. В каждый компонент добавить уникальный class-name на самый верхний уровень (plasma-button, plasma-badge) и т.д. (подумать на каком уровне то нужно сделать - ядро или библиотека)
3. Для каждого компонента из new-hope необходимо смапить токены для режимов onLight, onDark и inverse
4. Положить смапленные токены на уровень конфига  компонента
5. Считаем, что наши значения - эталонные, то есть пользователи должны будут учитывать эти соответствия при создании своих токенов и используя их в маппере
6. Автоматически сформировать список доступных компонент для типизации
7. Создать уникальные для каждой библиотеки (plasma-web / plasma-b2c) миксины applyOnLight / applyOnDark / applyInverse, в которых будут использоваться уникальные классы компонент

### Дополнительный функционал

- Реализовать компоненты `<LightContainer />` / `<DarkContainer />` / `<InverseContainer />` или `<ColoredContainer mode="onLight | onDark | inverse"/>` с зашитой внутри логикой по переопределению токенов цвета
- Добавить возможность использовать css-модули. То есть заранее сгенерировать файлы: `on-light.module.css`, `on-dark.module.css`, `inverse.module.css`, в которых будут лежать классы, которые пользователь сможет точечно добавлять на контейнеры.
 
Достоинства и недостатки метода: 
`+` Нет необходимости указывать для каждо компонента режим отображения. Т.к. частый кейс - это именно использования группы компонент в контейнере
`+` Так жесть возможность точечного указания компонент, которые необходимо модифицировать
`+` Работает в билд тайме, т.к. при сборке библиотеки сформируется только те css правила, которые были выбраны
 
`-` Необходимо произвести большую работу по ревизии всех используемых токенов в компонентах
`-` Возможно иногда пользователям понадобиться делать дополнительные врапперы, чтобы стилизовать набор компонент в контейнерах

**Comment by Yakutoc** [comment]
> `<ColoredContainer mode="onLight | onDark | inverse"/>`

Я за этот вариант.  

> Необходимо произвести большую работу по ревизии всех используемых токенов в компонентах

ИМХО это скорее плюс. 

- актуализируем знания
- ревизия кода

**Comment by neretin-trike** [comment]
> ИМХО это скорее плюс.

ну я тоже не до конца был уверен, что прям минус. Прост, объём времени достаточный понадобится для этого. Но тут скорее при любом подходе такое нужно будет делать

**Comment by TitanKuzmich** [comment]
Я правильно понимаю, что будет кастомный миксин, в котором вшит набор токенов по каждому компоненту. Этот миксин будет применятся в стилях контейнера LightContainer, к примеру?

**Comment by neretin-trike** [comment]
да, либо миксин можно использовать на своем контейнере, либо наш компонент.

Кажется, что проблем не должно быть, так как токены изолируются на верхних уровнях контейнеров если их друг в друга не вкладывать.

А вот если вкладывать, то надо понять, хотим ли мы еще и такой кейс поддерживать

**Comment by Yeti-or** [comment]
@neretin-trike это же реализовано в viewContainer ?
